### PR TITLE
Ticket 50052

### PR DESCRIPTION
**Pull request by azaozz**
Trac ticket: https://core.trac.wordpress.org/ticket/50052

This is an extension/additional PR to @whyisjake branch: https://github.com/whyisjake/wordpress-develop/tree/auto-updates-50052 that contains the core patch for "Plugin and Theme Auto-Updates".

### CODE REVIEW

File: src/wp-admin/network/themes.php [hunk 1]
```diff
@@ -208,6 +216,94 @@
...
+			check_admin_referer( 'bulk-themes' );
+
+			$themes = isset( $_POST['checked'] ) ? (array) wp_unslash( $_POST['checked'] ) : array();
+
+			$all_items    = wp_get_themes();
+			$auto_updates = (array) get_site_option( 'auto_update_themes', array() );
+
+			$new_auto_updates = array_diff( $auto_updates, $themes );
+
+			if ( $new_auto_updates == $auto_updates ) {
```
Comment (pbiron) [inline src/wp-admin/network/themes.php:291]: Following up from https://github.com/WordPress/wordpress-develop/pull/280/files#r426991577

For both bulk cases, there should be

```
if ( empty( $themes ) ) {
	wp_safe_redirect( add_query_arg( 'error', 'none', $referer ) );
	exit;
}
```

and then just `	wp_safe_redirect( add_query_arg( '{enabled|disabled}-auto-update', count( $themes ), $referer ) );` at the end

File: src/wp-admin/network/themes.php [hunk 2]
```diff
@@ -309,10 +405,36 @@
 		$message = _n( '%s theme deleted.', '%s themes deleted.', $deleted );
 	}
 	echo '<div id="message" class="updated notice is-dismissible"><p>' . sprintf( $message, number_format_i18n( $deleted ) ) . '</p></div>';
+} elseif ( isset( $_GET['enabled-auto-update'] ) ) {
+	$enabled = absint( $_GET['enabled-auto-update'] );
+	if ( 1 == $enabled ) {
```
Comment (pbiron) [inline src/wp-admin/network/themes.php:410]: Following up from https://github.com/WordPress/wordpress-develop/pull/279#discussion_r426841452

This should be `if ( 1 === $enabled ) {}` and the same for `$disabled` below

File: src/wp-admin/includes/class-wp-automatic-updater.php [hunk 3]
```diff
@@ -854,6 +862,190 @@ protected function send_email( $type, $core_update, $result = null ) {
...
+				break;
+		}
+
+		// Get failed plugin updates.
+		if ( in_array( $type, array( 'fail', 'mixed' ), true ) && ! empty( $failed_updates['plugin'] ) ) {
+			$body[] = __( 'The following plugins failed to update:' );
+			// List failed updates.
+			foreach ( $failed_updates['plugin'] as $item ) {
+				/* translators: %s: Name of the related plugin. */
+				$body[] = ' ' . sprintf( __( '- %s' ), $item->name );
```
Comment (audrasjb) [inline src/wp-admin/includes/class-wp-automatic-updater.php:980]: This line can be replaced with:
`$body[] = "- {$item->name}";`

For reference: https://github.com/WordPress/wp-autoupdates/pull/138/files

File: src/wp-admin/includes/class-wp-automatic-updater.php [hunk 4]
```diff
@@ -854,6 +862,190 @@ protected function send_email( $type, $core_update, $result = null ) {
...
+			}
+			$body[] = "\n";
+		}
+		// Get failed theme updates.
+		if ( in_array( $type, array( 'fail', 'mixed' ), true ) && ! empty( $failed_updates['theme'] ) ) {
+			$body[] = __( 'The following themes failed to update:' );
+			// List failed updates.
+			foreach ( $failed_updates['theme'] as $item ) {
+				/* translators: %s: Name of the related plugin. */
+				$body[] = ' ' . sprintf( __( '- %s' ), $item->name );
```
Comment (audrasjb) [inline src/wp-admin/includes/class-wp-automatic-updater.php:990]: This line can be replaced with:
`$body[] = "- {$item->name}";`

For reference: https://github.com/WordPress/wp-autoupdates/pull/138/files

File: src/wp-admin/includes/class-wp-automatic-updater.php [hunk 5]
```diff
@@ -854,6 +862,190 @@ protected function send_email( $type, $core_update, $result = null ) {
...
+			}
+			$body[] = "\n";
+		}
+		// Get successful plugin updates.
+		if ( in_array( $type, array( 'success', 'mixed' ), true ) && ! empty( $successful_updates['plugin'] ) ) {
+			$body[] = __( 'The following plugins were successfully updated:' );
+			// List successful updates.
+			foreach ( $successful_updates['plugin'] as $item ) {
+				/* translators: %s: Name of the related plugin. */
+				$body[] = ' ' . sprintf( __( '- %s' ), $item->name );
```
Comment (audrasjb) [inline src/wp-admin/includes/class-wp-automatic-updater.php:1000]: This line can be replaced with:
`$body[] = "- {$item->name}";`

For reference: https://github.com/WordPress/wp-autoupdates/pull/138/files

File: src/wp-admin/includes/class-wp-automatic-updater.php [hunk 6]
```diff
@@ -854,6 +862,190 @@ protected function send_email( $type, $core_update, $result = null ) {
...
+			}
+			$body[] = "\n";
+		}
+		// Get successful theme updates.
+		if ( in_array( $type, array( 'success', 'mixed' ), true ) && ! empty( $successful_updates['theme'] ) ) {
+			$body[] = __( 'The following themes were successfully updated:' );
+			// List successful updates.
+			foreach ( $successful_updates['theme'] as $item ) {
+				/* translators: %s: Name of the related plugin. */
+				$body[] = ' ' . sprintf( __( '- %s' ), $item->name );
```
Comment (audrasjb) [inline src/wp-admin/includes/class-wp-automatic-updater.php:1010]: This line can be replaced with:
`$body[] = "- {$item->name}";`

For reference: https://github.com/WordPress/wp-autoupdates/pull/138/files

File: src/wp-admin/includes/update.php [hunk 7]
```diff
@@ -933,3 +938,76 @@ function wp_recovery_mode_nag() {
...
+			 * Filters whether plugins manual auto-update is enabled.
+			 *
+			 * @since 5.5.0
+			 *
+			 * @param bool True if themes auto-update is enabled, false otherwise.
+			 */
+			return apply_filters( 'wp_themes_auto_update_enabled', true );
+	}
+
+	return false;
```
Comment (pbiron) [inline src/wp-admin/includes/update.php:972]: Following up on https://github.com/WordPress/wordpress-develop/commit/86941f9f0979e9b879c38fa7e85410b9bedf1a2f#r39289435

Would be good to add

```
	case 'translation':
		return true;
```

to the switch.

Comment (azaozz) [reply src/wp-admin/includes/update.php:972]: Not sure this is a good idea.
- The feature plugin is for auto-updating plugins and themes, there's nothing in it about translations.
- There is no UI for setting specific translation files to be auto-updated by the users.
- There is no option `auto_update_translations` (that would be needed/checked if `wp_is_auto_update_enabled_for_type( 'translation' )` returns true.

File: src/wp-admin/includes/class-wp-automatic-updater.php (hunk 6 above)
Comment (audrasjb) [reply src/wp-admin/includes/class-wp-automatic-updater.php]: Thanks for the update 👍

**Comment by whyisjake** [comment]
Merged!

**Pull request closed without merging**